A11Y Improvements -- Render buttons as buttons

jodylecompte opened the pull request:
Per #148, I have been playing with the application tonight and the first thing that leapt out to me was that the form on the "New Project" modal was not able to be submitted by keyboard. The root cause of this that this PR fixes is rendering buttons as buttons rather than a div/span. 

This can cause multiple accessibility challenges both to keyboard navigation and control and also screen reader identification. Forms throughout the application are now able to be submitted by keyboard so long as they utilize a real `<form>` element and an `onSubmit` property.

midnightprioriem commented on app/renderer/controls/Button.tsx:
```diff
@@ -9,7 +9,7 @@ type StyledButtonProps = {
 };
 
-const StyledButton = styled.span<StyledButtonProps>`
+const StyledButton = styled.button<StyledButtonProps>`
```
Can you add style for the focus outline? I have defined a new focused style for the button:

![State=Focused](https://user-images.githubusercontent.com/6120081/193588130-7803ca1a-fc9e-4b4d-8ee9-86de032ba8e0.png)

It is a 4px solid outline using theme.buttonPrimaryBg at 50% alpha. You can use `Color(theme.buttonPrimaryBg).alpha(0.5).toString()` to acheive this.

Be sure to use the `focus-visible` psuedo-class instead of just `focus` to prevent the outline from being shown when interacting with the button via mouse.

jodylecompte replied:
Certainly! The requested change has been added to this PR.

midnightprioriem reviewed (changes requested):
Thanks for this change! I just had one minor comment to address styling that will likely carry forward into future a11y changes.

midnightprioriem commented on app/renderer/controls/Button.tsx:
```diff
@@ -19,12 +19,17 @@ const StyledButton = styled.span<StyledButtonProps>`
 
+  &:focus-visible {
+    outline: ${(p) =>
+      `4px solid ${Color(p.theme.buttonPrimaryBg).alpha(0.5).toString()};`}
```
I think there's a missing curly brace here! Also, you could write the outline as:
```css
outline: 4px solid
      ${(p) => Color(p.theme.buttonPrimaryBg).alpha(0.5).toString()};
```

No need for the template literals.

jodylecompte replied:
Yikes, I'm not sure what I was thinking when I put that template literals there. 😬 

I have corrected that, removed the onClick handler, and cleaned up a few other small details (import order, curly brackets around string literal) that ESLint was complaining about.

midnightprioriem commented on app/renderer/modals/NewBookModal.tsx:
```diff
@@ -20,6 +20,7 @@ const NewBookModal = (props: ModalProps) => {
   const formRef = useRef<HTMLFormElement>(null);
   const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
```
I don't think the Button in the form doesn't need the `onClick` prop to be set anymore, so it should be removed to prevent the form from getting submitted twice.

midnightprioriem reviewed (changes requested):
Sorry, just noticed one more minor thing!

We should be passing `disabled={disabled}` to `StyledButton` so that form submissions with the button are not possible if we've disabled the button. Minor thing, but it allows users to use the Generate Book button even if it's disabled!

jodylecompte commented:
> Sorry, just noticed one more minor thing!
> 
> We should be passing `disabled={disabled}` to `StyledButton` so that form submissions with the button are not possible if we've disabled the button. Minor thing, but it currently we can use Generate Book button even if it's disabled!

No problem, but I want to make sure I am looking at the right thing. In `NewBookModel` there does not seem to be disabled value or any sort of toggle, but in `GenerateBookModel`, the prop `isDisabled` is being passed in and then used on `StyledButton`, but I'm assuming this needs to be changed from `isDisabled` to `disabled` to match the correct prop for a standard HTML button?

midnightprioriem commented:
> No problem, but I want to make sure I am looking at the right thing. In `NewBookModel` there does not seem to be disabled value or any sort of toggle, but in `GenerateBookModel`, the prop `isDisabled` is being passed in and then used on `StyledButton`, but I'm assuming this needs to be changed from `isDisabled` to `disabled` to match the correct prop for a standard HTML button?

Sorry, the GitHub reviewer doesn't let you make comments on _any_ line in the code yet.

Yep, the change should be made in `Button.tsx` and should be exactly what you said, pass the disabled prop down to the standard HTML button. I think you're right, you can also remove the `isDisabled` prop from StyledButtonProps and use `p.disabled` directly in the css.

jodylecompte commented:
I think this is good to go now. I changed it to `disabled` on the button itself, but maintained `isDisabled` in higher elements to create distinction between the abstraction and the element, but I can make it `disabled` all the way up if you would prefer that. 

I also noticed in testing that the button maintained the click cursor if the button has been disabled, so I fixed that also.

midnightprioriem commented on app/renderer/controls/Button.tsx:
```diff
@@ -19,12 +19,22 @@ const StyledButton = styled.span<StyledButtonProps>`
   text-overflow: ellipsis;
   font-size: 0.9em;
 
+  &:disabled {
```
Instead of this I think you just need to add `p.disabled` to line 41 like so:

```css
${(p) => 
    !p.isLoading &&
    !p.disabled &&
    css`
```

This will take care of the pointer and background-color styling that is needed for when the button is disabled. And the prop on line 9 can be removed since `isDisabled` is not being passed to `StyledButton` anymore.

midnightprioriem commented on app/renderer/modals/GenerateBookModal.tsx:
```diff
@@ -66,20 +66,14 @@ const GenerateBookModal = (props: ModalProps) => {
-            }}
-            disabled={!pdfPlatformToggleValue}
-          >
+          <Button loading={isBuildingPdf} disabled={!pdfPlatformToggleValue}>
```
We should set `isDisabled` here now since the Button prop changed.

midnightprioriem commented:
Looking good, these should be the last few comments!

jodylecompte commented:
I think we are good to go now, sorry you've had to go over it so many times!

midnightprioriem reviewed (approved):
Everything looks great now!
